### PR TITLE
fix(47383): Destructuring of unknown catch variable is not an error

### LINKED ISSUE #47383: Destructuring of `unknown` catch variable is not an error
# Bug Report

### 🔎 Search Terms

catch, destructure, destructuring, useUnknownInCatchVariables

### 🕗 Version & Regression Information

- This is the behavior in every version I tried: v4.0.5, v4.5.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.4#code/DYUwLgBAhgXBCuA7A1og9gd0QbgFCkgG8QAnEtEgXwgF5o9cAzJAYzAEs1EIAjKEgBSFGaNJThJUmRAEoIhSrlxgSAT3m5qLKGBYALCENLkqElOixzCmoA)

### 💻 Code

```ts
let a: unknown;
let {foo} = a;
//   ^^^ Property 'foo' does not exist on type 'unknown'.(2339)

function bar({foo}: unknown) {}
//            ^^^ Property 'foo' does not exist on type 'unknown'.(2339)

try {
} catch ({foo}: unknown) {
  // foo is type any
  // with useUnknownInCatchVariables foo is type unknown
}
```

### 🙁 Actual behavior

Destructuring `unknown` is an error, but destructuring a catch variable, which is of type `unknown` is not an error.

### 🙂 Expected behavior

I would expect destructuring of a catch variable of type `unknown` to produce an error.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #47383

### CODE REVIEW

File: tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt
Comment (RyanCavanaugh) [inline tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt:11]: This seems like a weird error -- where is `{}` coming from?

Comment (a-tarasyuk) [reply tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt:11]: @RyanCavanaugh I think there are two ways to throw the error, use `TS2339` directly in condition with `useUnknownInCatchVariables`. or use `getBindingElementTypeFromParentType` similarly to parameter binding element which actually has error reporting `TS2339` and uses any type instead of `unknown`.

```ts
function bar({ foo }: unknown) {
   foo // any
}
```

Comment (jakebailey) [reply tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt:11]: I'd think the latter, so everything is consistent with destructuring `unknown` elsewhere. [Playground Link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAbgLhgVzAazCA7mGBeGARgCYBmAFgG4AoW0SWAbxgDMQQYBffeGgej4whwmAD0A-EA)

Comment (jakebailey) [reply tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt:11]: I just debugged this and realized that this is the `strict=false` test; the `strict=true` test shows the desired error (and the code is doing the "latter" I said above).

This seems to be because `getApparentType` returns the empty object for `unknown` when `strictNullChecks` is disabled.

Comment (jakebailey) [reply tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt:11]: And further, `getPropertyTypeForIndexType` uses that reduced apparent type for its error messages, not the original type.

Comment (jakebailey) [reply tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt:11]: Surprisingly, this is already the behavior, it's just that there's no test for it!

image

Comment (jakebailey) [reply tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt:11]: For fun, I removed the conversion of `unknown` to `{}`, and honestly, I can't say I'm a fan of other existing errors like:

```
 !!! error TS2322:  Type 'unknown' is not assignable to type 'A'.
 !!! error TS2322:    Type 'ReturnType<FunctionsObj<T>[string]>' is not assignable to type 'A'.
 !!! error TS2322:       Type 'unknown' is not assignable to type 'A'.
 !!! error TS2322:         Property 'x' is missing in type '{}' but required in type 'A'.
```

I mean, obviously we know that `unknown` isn't assignable to anything, so what's the point in picking out a prop of `A` to say that it's missing that prop?

Comment (jakebailey) [reply tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt:11]: I filed #47494 to cover this.

**Comment by jakebailey** [comment]
This PR does pass tests, but I'm a bit confused as to why it doesn't work when I run the build via tsserver locally in VS Code; do these sorts of checks need to be somewhere else, maybe? Some missing case in another function?

File: tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt
Comment (jakebailey) [reply tests/baselines/reference/useUnknownInCatchVariables02(strict=false).errors.txt:11]: So, in the context of this PR, I don't think that this actually is anything new and is okay to go in; we can use the above issue to talk about changing the apparent type of `unknown` (which would help this and other cases, but maybe break things).

**Comment by jakebailey** [comment]
Hm, apparently `checkTryStatement` doesn't actually call a check function on the catch clause / its variable declaration. That doesn't seem right.

**Comment by jakebailey** [comment]
Yeah, I think `checkTryStatement` needs `checkVariableLikeDeclaration(declaration)`, but that seems to cause some baseline changes including follow-on errors about implicit `any`, which are probably not right.

**Review by jakebailey — CHANGES_REQUESTED** [review]
Marking as needing changes (meant to do this with my last comment), as this doesn't seem to work in VS Code (and probably other editors), so we'll need to make sure that check call is added. Not entirely certain how to ensure the tests actually walk that; maybe they already check the opposite?

**Comment by a-tarasyuk** [comment]
@jakebailey Thanks for the review and sorry for the delay in response. I've added changes that cover the following cases

```ts
// @strict: true
// @useUnknownInCatchVariables: true

try {
    // ...
}
catch ({ name }) {
         ^^^^ Property 'name' does not exist on type 'unknown'
         // { name } has 'unknown' type because @useUnknownInCatchVariables: true
    name; 
}

try {
    // ...
}
catch ({ name } : unknown) {
         ^^^^ Property 'name' does not exist on type 'unknown'
         // { name } has explicit type 'unknown'
    name;
}

```

```ts
// @strict: true
// @useUnknownInCatchVariables: false

try {
    // ...
}
catch ({ name }) {
        // Ok
        // { name } has 'any' type because @useUnknownInCatchVariables: false
    name; 
}

try {
    // ...
}
catch ({ name } : unknown) {
         ^^^^ Property 'name' does not exist on type 'unknown'
         // { name } has explicit type 'unknown'
    name;
}

```

```ts
// @strict: false
// @useUnknownInCatchVariables: true

try {
    // ...
}
catch ({ name }) {
         ^^^^ Property 'name' does not exist on type '{}' 
         // { name } has '{}' type because @useUnknownInCatchVariables: true
         // `unknown` is represented as `{}` with disabled `strict: false` check
    name; 
}

try {
    // ...
}
catch ({ name } : unknown) {
         ^^^^ Property 'name' does not exist on type `{}`
         // { name } has explicit type 'unknown'
    name;
}

```

```ts
// @strict: false
// @useUnknownInCatchVariables: true

try {
    // ...
}
catch ({ name }) {
         // Ok
         // { name } has 'any' type because @useUnknownInCatchVariables: false
    name; 
}

try {
    // ...
}
catch ({ name } : unknown) {
         ^^^^ Property 'name' does not exist on type `{}`
         // { name } has explicit type 'unknown'
    name;
}
```

Is this the expected behavior? Or do we need to cover other cases?

**Comment by jakebailey** [comment]
Unless I'm mistaken, all of the examples above should have `name` be of type `any`, and that's what the PR has (so maybe your comment is wrong), so that's still good and consistent with other restructuring.

It looks like things are still not quite right in the editor, though. If you use your test case but don't reference the variable in question, it still doesn't show an error. Compare the three cases here, each with the destructured variable left referenced.

```ts
const { xyz }: unknown = {};

function foo({ name }: unknown) {
    // ...
}

try {
    // ...
}
catch ({ name }: unknown) {
    // ...
}
```

image


I'm not sure what you've changed to get further is correct, though; it still feels to me like there's something missing during the checker walk, and these errors are only functioning due to the side effect of evaluating the variable later.

File: src/compiler/checker.ts
Comment (jakebailey) [inline src/compiler/checker.ts:38336]: I think it'd be better to stick `checkVariableLikeDeclaration(declaration)` right below its declaration above; I don't know if there's predicant for conditionally walking like this based on type.

Doing that `checkVariableLikeDeclaration` does introduce a load of "implicit any" errors, which I _think_ are arguably correct, but would probably be a sweeping change.

Pending that, I don't think that this `forEach` is correct; it seems like it doesn't recurse down when it should for consistency (see `destructuringCatch`'s examples with `catch ([{x: [y], z}])` and such). That test also poses an interesting question about the behavior of something like `catch ([a, b])`, though.

Comment (a-tarasyuk) [reply src/compiler/checker.ts:38336]: @jakebailey I've converted the PR into a project. I'll check it later when I have time.

Comment (jakebailey) [reply src/compiler/checker.ts:38336]: No worries; I really appreciate you working on this.

Just to note it here before I forget, if I'm remembering correctly, errors about the "same thing" are to be avoided; [my example](https://www.typescriptlang.org/play?#code/MYewdgzgLgBA3jAHgTwF4wL4C4YFcwDWYIA7mDALzwYDcAUKJCADYCmAdMyAOYAUKqAJT06AM3zAoAS3AxRIELwRgAhgFtWmHPiKkwg+HRjGYjCCw5c+qjcLoY6dKACdkhkzAD0nmOz-2GFShgAAsYJRgbTWw8QmIyAzgjEzMLTh5eKLsMIA) shows that the existing destructuring of `unknown` don't issue `implict any`, so there's something to look into here about how those are walked and how they avoid that double error.

File: tests/baselines/reference/catchClauseWithTypeAnnotation.errors.txt
Comment (jakebailey) [inline tests/baselines/reference/catchClauseWithTypeAnnotation.errors.txt:5]: Interesting; this seems to imply that this PR now enables annotations of non any/unknown type (though, we have the explicit error for it). Probably out of scope for this particular fix though, but I do think that's interesting.

Comment (a-tarasyuk) [reply tests/baselines/reference/catchClauseWithTypeAnnotation.errors.txt]: I reverted to `any` | `unknown`. Yes, this is out of scope, I was just experimenting with this behavior.

File: src/compiler/checker.ts
Comment (jakebailey) [inline src/compiler/checker.ts:38336]: I think this should be `isErrorType` (looking at the rest of the code).

File: tests/baselines/reference/redeclareParameterInCatchBlock.symbols
Comment (jakebailey) [inline tests/baselines/reference/redeclareParameterInCatchBlock.symbols:34]: This seems suspicious.

File: tests/baselines/reference/destructuringCatch.symbols
Comment (jakebailey) [inline tests/baselines/reference/destructuringCatch.symbols:34]: This one too; these have symbols in other destructurings.

File: src/compiler/checker.ts
Comment (jakebailey) [reply src/compiler/checker.ts:38336]: However, I'm a bit confused how this works after the change; before this read as "if it's not any or unknown, error", which is the spirit of the diagnostic.

**Comment by sandersn** [comment]
I'm confused about the state of this PR. Why is it marked draft? If a team member signs off, is it OK to merge? If so, I guess that means it needs another review?

**Comment by a-tarasyuk** [comment]
@sandersn Sorry for the confusion with the state of status with PR. I've closed it to make this issue open to PR acceptance.

**Pull request closed without merging**
